daemon/src/engine/server/CryptoChallege.h

```diff
@@ -0,0 +1,633 @@
+    stream.fill('0');
+    for ( auto ch : input )
+    {
+        stream.width(2);
+        stream << int( ch );
+    }
+    return stream.str();
+}
+
+/*
+ * Translates a hexadecimal string into binary data
+ * Throws Crypto::Error if the data is not a valid hexadecimal string
+ */
+inline Data HexDecode( const std::string& input )
+{
+    if ( input.size() % 2 )
+    {
+        throw Error( "Invalid Hex string" );
+    }
+
+    Data output( input.size() / 2 );
+
+    for ( std::size_t i = 0; i < input.size(); i += 2 )
+    {
+        if ( !Str::cisxdigit( input[i] ) || !Str::cisxdigit( input[i+1] ) )
+        {
+            throw Error( "Invalid Hex string" );
+        }
+
+        output[ i / 2 ] = ( Str::gethex( input[i] ) << 4 ) | Str::gethex( input[i+1] );
+    }
+
+    return output;
+}
+
+/*
+ * Translates binary data into a base64 encoded string
+ */
+inline std::string Base64Encode( const Data& input )
+{
+    base64_encode_ctx ctx;
+    nettle_base64_encode_init( &ctx );
+
+    Data output( BASE64_ENCODE_LENGTH( input.size() ) + BASE64_ENCODE_FINAL_LENGTH );
+    int encoded_bytes = nettle_base64_encode_update(
+        &ctx, output.data(), input.size(), input.data()
+    );
+    encoded_bytes += nettle_base64_encode_final( &ctx, output.data() + encoded_bytes );
+    output.erase( encoded_bytes );
+    return output.str();
+}
+
+/*
+ * Translates a base64 encoded string into binary data
+ * Throws Crypto::Error if the data is not a valid hexadecimal string
+ */
+inline Data Base64Decode( const Data& input )
+{
+    base64_decode_ctx ctx;
+    nettle_base64_decode_init( &ctx );
+
+    Data output( BASE64_DECODE_LENGTH( input.size() ) );
+    std::size_t decoded_bytes = 0;
+    if ( !nettle_base64_decode_update( &ctx, &decoded_bytes, output.data(),
+                                       input.size(), input.data() ) )
+    {
+        throw Error( "Invalid Base64 string" );
+    }
+
+    if ( !nettle_base64_decode_final( &ctx ) )
+    {
+        throw Error( "Invalid Base64 string" );
+    }
+
+    output.erase( decoded_bytes );
+    return output;
+}
+
+
+} // namespace Encoding
+
+
+namespace Hash {
+
+inline Data Sha256( const Data& input )
+{
+    sha256_ctx ctx;
+    nettle_sha256_init( &ctx );
+    nettle_sha256_update( &ctx, input.size(), input.data() );
+    Data output( SHA256_DIGEST_SIZE  );
+    nettle_sha256_digest( &ctx, SHA256_DIGEST_SIZE, output.data() );
+    return output;
+}
+
+inline Data Md5( const Data& input )
+{
+    md5_ctx ctx;
+    nettle_md5_init( &ctx );
+    nettle_md5_update( &ctx, input.size(), input.data() );
+    Data output( MD5_DIGEST_SIZE );
+    nettle_md5_digest( &ctx, MD5_DIGEST_SIZE, output.data() );
+    return output;
+}
+
+
+} // namespace Hash
+
+namespace Encryption {
+
+/*
+ * Adds PKCS#7 padding to the data
+ */
+inline void AddPadding( Data& target, std::size_t block_size = 8 )
+{
+    auto pad = block_size - target.size() % block_size;
+    target.resize( target.size() + pad, pad );
```

Comment:
This function follows a standard specification. It's the standard way of padding data for encrypted stuff.